### ISSUE TITLE
Allow ember-mirage to generate the serializers (primaryKey and attrs) like it does for models 👍

### ISSUE THREAD

**Issue by cah-briangantzler** [miragejs/ember-cli-mirage#1989]
I have a PR in the works, we have been using it at my company for a while now, wanted to get your take on it, any suggestions etc.  While fully usable, there are two issues that I cant seem to figure out and was hoping we could pair or something to get through them.   Those issues will be detailed at the bottom:

My current fork:  https://github.com/bgantzler/ember-cli-mirage/tree/serializer
Expect some of these commits to be rebased or altered as I continue testing, as a PR has not yet been submitted.

Create a new serializer based off the RestSerializer called EmberDataSerializer that does the following:

- Altered the Server to generate mirage serializers based off the primaryKey and attrs hash (will be called transforms in the mirage serializer (attrs was already in use)) defined in ember data serializers. (If a mirage serializer already exists, it will extend them)
- both serialize and normalize(deserialize) will transform to and from the id property of the mirage store using the specified primaryKey
- both serialize and normalize(deserialize) will use a new hash `transforms` as follows:
- allows each property to be renamed in the JSON payload like the `key` property in the attrs hash for ember data serializers.
- allows for each relation to specify how serialize and normalize will treat the JSON payload . Values are the same as the ember data serializer. Includes the value `records` from the embeddedRecordsMixin. The serialize from the ember data serializer will be applied to the mirage deserialize, and the deserialize of the ember data serializer will be applied to the serialize of the mirage serialize.
- still respects the includes array for side loading additional elements.
- does not respect the mirage attrs array (uses the transforms hash to specify which properties to include)
- does not respect alwaysSerializeIds (uses the transforms to allow each relationship to specify if ids should be included)
- does not respect embed (uses transforms hash to allow each relationship to specify if it should be embedded)
- Includes the pushMirageToStore method. This method works the same as the one documented in issues with the following changes. Works the same if no parameters are passed. Allows for a hash to be passed. This hash will allow you to select which tables to push and an optional where clause to be applied to that table.
- For addons, provides a ember-cli-build method that can be imported into the index.js that provides for exporting the mirage folder into the application.

The two issues I am currently running into do not stop the use of these changes, just causes two minor annoyances

1) when I create an application serializer, and specify `root: false` it does not seem to be respected in the serializer code. I have to do the following, I dont understand why 

```
normalize() {
    this.root = false;
    return EmberDataSerializer.prototype.normalize.apply(this, arguments);
}
```

2) when the exportToMirage is imported from ember-cli-mirage repo, the merged trees are at the wrong level. When the code is copied into the addon repo, then imported, it is correct. It Is the same code in both cases, the code is using locally provided arguments, I dont understand why it would make a difference where it is being imported from.

**Comment by samselikoff** [comment]
Awesome.

After we ship the tutorial + v1.0 of miragejs and get back to feature development, I'd like to focus on what primitives are missing from Mirage that would allow a solution like this to exist in userland. We are thinking through a plugin or middleware system.

As you've been building this, what changes or new primitives would we need to expose so you could build this completely in userland?

**Comment by cah-briangantzler** [comment]
What do you mean by userland, do you consider ember-cli-mirage as user land for ember?

The real change was that the current flags in mirage were an all or nothing, being allied to all properties in the serializer, so I couldn't reverse engineer the an ember data serializer into them, so I created the transforms hash (equiv of the attrs hash in ED Serializer). That hash also allowed for a property by property rename. 

I kept that in ember-cli-mirage instead of making a mirageJS PR because it so closely mirrored ED way of doing it and didnt think you wanted two ways of doing things. But because it gives a much finer control the transforms hash could easily be implemented in the miragesJS rest (maybe active) serializer. I dont think think it would apply to JsonAPI.

The other major change was to the StartMirage code to create the serializers from ED serializers. This would only apply to user land of ember, which is what this add on is all about.

The bulk of the code was the transforms hash, and allowing for defaults. The renaming used the keyForAttribute. Not having id go through there and instead having a keyForId would have been nicer, but only a couple lines of code there.

Most the rest of the code was in _hashForModel and a shouldSerializeProperty would have allowed the include/exclude and could be implemented by the whitelist attrs, then maybe a serialzeProperty (have some ideas there, in fact may change my code to implement this, then that could be a mirageJS PR).

**Comment by samselikoff** [comment]
By userland I meant in your application itself, rather than having to submit a PR to `miragejs` or `ember-cli-mirage`.

I was curious if the changes you're proposing require changes to the core lib itself, or if it's more just a new serializer that would be convenient for Ember Data users to use.

**Comment by samselikoff** [comment]
Also this PR is big. We should break it down into the smallest pieces to maximize the chance of shipping it. There's been lots of large PRs that have languished in Mirage's history.

For example the pushStoreIntoDb is kind of a separate problem. It needs addressing, but probably not in this PR.

**Comment by cah-briangantzler** [comment]
Effectively yes, this is a new serializer that is more specific to ember data that I think belongs in ember-cli-mirage as everyone should not have to reimplement this.

The push to store and merge trees could easily be moved out for a separate pr, they are really each a stand alone files. So drop those two files.

While the serializer could be in its own PR, its value is gained by the change to the startup which like models creates mirage serializers from the ED serializers, so I would prefer that remains as a whole.

**Comment by samselikoff** [comment]
Gotcha.

I wonder if we should do this as a separate package that folks can install, so you can experiment without me blocking you. My goal is eventually for `ember-cli-mirage` to be a very small wrapper around `miragejs` and this kind of moves it in the opposite direction. That's also the reason I asked whether this could be done in "userland", meaning could it be done in a separate library without any new changes or APIs added to `miragejs`.

Also, different apps that use Ember Data can still have different API formats. Is the idea with this serializer that the API format information is already contained in the Ember Data adapter/serializer layer, so we can just leverage that and save us the work of defining it on the Mirage side?

**Comment by cah-briangantzler** [comment]
Correct

A separate package for the serializer would work for the serializer alone, but the real key is like models, creating the serializers off the ED serializers, so I would have to subclass startMirage, and then server, and effectively would be a replacement to ember-cli-mirage.

**Comment by cah-briangantzler** [comment]
Ember-cli-mirage would be a thin wrapper to MirageJS except for the startMirage and Mirage doing the model generation, you still need to keep that code. I want to add to that code by generating serializers.    Exposing some more primitives in mirageJS might reduce some of the code here, but they way it renames fields, specified embed vs ids, is different than what MirageJS does, so Im not sure that would make it into MirageJS

**Comment by samselikoff** [comment]
Gotcha. Ya so that's exactly what I was talking about, what parts of mirage or ec-mirage do we need to expose so you could build something like this in a separate package without needing to change mirage.

I think ideally, ec-mirage exposes a function that gets the Ember Data model and serializer map passed into it as arguments, and you can use those to help you define any mirage config for your app.

**Comment by samselikoff** [comment]
Yes totally – the startMirage and auto discovery of models is an ec-mirage concern. I think it would be good to put some better boundaries around this code and the code that configures the Mirage server.

**Comment by cah-briangantzler** [comment]
So what you would need to expose would be some way for another addon to hook into startMirage and Server. I dont have a suggestion for that.

**Comment by samselikoff** [comment]
Server has a `config()` method that can take it additional configuration, so you might be able to do it in `config.js` like this:

```js
// config.js
export default function() {

  // Add additional serializers
  this.config({
    serializers: {
      application: SomeSerializer
    }
  })
}
```

So in theory an addon could expose a function that the host app could use to augment their server with the right serializers:

```js
// config.js
import { defineSerializers } from 'ember-cli-mirage-ember-data'

export default function() {

  // Add additional serializers
  defineSerializers(this, mySerializers)
}
```

So the next question is how best to make the application's serializers (or more generally, the app's Ember Data config) available to either this function or to the `config.js` Mirage file.

We could do it by exporting a new named function from `config.js` that gets the config and lets users update it. And we could inject the ember data information into this function so users could use this to define their mirage config based on Ember Data config. And then the autodiscovery of ember data models features could be implemented in terms of this API.

Something along the lines of

```js
// config.js
import { defineSerializers } from 'ember-cli-mirage-ember-data'

export function augmentConfig({ config, emberData }) {
  config.serializers = defineSerializers(emberData.serializerMap)
}

export default function() {
  // Normal Mirage config
}
```

**Comment by cah-briangantzler** [comment]
The first option, this.config in the config.js wont work, this function is called too late, if you implemented models there the ED models would override the mirage models. In the case of serializers we dont want to override but extend.

The second if the most promising, if that method would be called before the schema is created here https://github.com/miragejs/miragejs/blob/master/lib/server.js#L464 it would allow the function to add, remove, alter the models and serializers. If it was called earlier, like here https://github.com/miragejs/miragejs/blob/master/lib/server.js#L237 we would be able to augment more, timing, logging, trackRequest, etc.

The parameter emberData would contain what? serializerMap and  modelMap? or should that not be passed and its the addons problem to do https://github.com/miragejs/ember-cli-mirage/blob/master/addon/ember-data.js.  Presumably this would be moved to the addon that implemented model discovery. I basically copied this code for and changed the regex for the serializer discovery, this wouldnt need to be part of ec-mirage anymore but part of the ec-mirage-model addon.

The nice thing about embers  addons are install and they pretty much just work. For a user to use ec-mirage now, the config would have to look like this

```
// config.js
import { getModels } from 'ember-cli-mirage-ember-models'; // assuming model discovery moved from ec-mirage to its own addon

export function augmentConfig({ config, emberData }) {
  options.models = getModels(config.models); // change to allow param to keep Object.assign in the method
}

export default function() {
  // Normal Mirage config
}
```
a breaking change. But it is clear where the user is opting into model discovery, so the config autoDiscoverModels could go away


and then to use my addon they would change to 
```
import { getModels } from 'ember-cli-mirage-ember-models'; 
import { getSerializers } from 'ember-cli-mirage-ember-serializers'; 

export function augmentConfig({ config, emberData }) {
  options.models = getModels(config.models); 
  options.serializers = getSerializers(config.serializers)
}

export default function() {
  // Normal Mirage config
}
```

This leaves ec-mirage as only a wrapper to start the server from startMirage and nothing else.

But do we really need two addons, ec-mirage and another to do the model discovery. And then a third for serializers. If you keep the model discovery in ec-mirage, well serializers go hand in hand with models so why not put that in ec-mirage as well. By putting everything in ec-mirage there is no additional config and the new serializer is available by just changing serializer application inherits from.



In the end, if you allow for an augmentConfig method as described above, yes, I could provide the serializerDiscovery in a separate addon.

**Comment by samselikoff** [comment]
> The nice thing about embers addons are install and they pretty much just work

While true, that's also why you've been blocked on this, because of the "magic" that ec-mirage uses to instantiate + configure the Mirage server. And it's not just you, there's a ton of open issues here related to the fact that Mirage starts itself in a magical way that is not obvious how to override.

I'd like to move away from adding more "magic", since it hides the lower-level pieces and makes it hard for folks like yourself to configure Mirage when you need to.

I think the `augmentConfig` export is too magical, it doesn't provide enough control. I think we should try to think of a way to expose more control in a backwards compatible way.

For example, we could detect a new file called `index.js` that lives in the `/mirage` directory, and if that file exists, we don't automatically define the factories/serializers/model definitions on the server. Instead, maybe this file exports a `createServer` function that has everything defined right there:

```js
// mirage/index.js
import { Server } from "miragejs";

export function createServer() {
  return new Server({
    models: {},
    serializers: {},
    // ....
  });
}
```

That way (1) it matches `miragejs` (2) users can see everything that's happening and they have the opportunity to wire up anything else they need. We'd also need some way to hook this into the initializer. So I think that's what we should focus on next, because I really don't want to add more ad-hoc APIs to ec-mirage since they just make the project harder to support.

We could eventually provide some utils to help wiring up the files on disk to Mirage but, I think a good starting place is to let folks import the files themselves and pass them in, since this is a lower-level API.

**Comment by cah-briangantzler** [comment]
I understand.  

Would makeServer.js be a better name? (I think there are many that create addons without understanding or even touching index.js)

Where are you going to draw the line on magic? You could remove setupMirage, startMirage and the Server defined in ec-mirage and just ask them to call this function in beforeEach, and then shutdown in afterEach. I think that would effectively eliminate ec-mirage as it is now. Then a new addon that does model discovery that they would import and use in this new makeServer function. 

I think that effectively what is happening here https://miragejs.com/quickstarts/cypress/#step-2-define-your-server.

I think the question is what is the vision of the functionality ec-mirage as an addon provides? Just the qunit wrapper startup (with the above suggestion to get a hook to the create server). Is the model discovery part of this addon? or another addon?

**Comment by samselikoff** [comment]
I think it would be totally reasonable for ec-mirage to provide the model & serializer discovery, and to either generate an initializer or provide a helper that does so. I just think the actual instantiation shouldn't be as hidden as it is today, so that users can see how the discovered models & serializers are passed in, and can therefore change it all.

Passing the Ember Data info into a startup function seems pretty cool, as a way to simplify the setup for the user but still give them the opportunity to change it.

Re: name, yeah we can definitely brainstorm. I wasn't thinking `index.js` like the addon, was just thinking something like `mirage/index.js` in folks' apps. We can definitely brainstorm though. Could also just have a new named export from `mirage/config.js`, so instead of the default function export we check to see if there's a named function export called something like `setupMirage` or `makeServer`.

**Comment by cah-briangantzler** [comment]
Obviously this would have been easier on a call or in person. :) I knew you meant mirage/index.js, I meant I think a lot of people dont realize what the name index.js means and might not look there or wonder why its called index.js (im fine with it, I understand it). 

Totally agree over how some of the magic happens by making it more clear. This also opens up the ability to get access to any new configurations on Server create from mirageJS.

Aside: how does https://github.com/miragejs/ember-cli-mirage/blob/master/app/initializers/ember-cli-mirage.js#L2 work when someone renames the mirage directory, since "mirage" is hardcoded on that import. Shouldnt that throw a module not found?

Since testConfig is already a special export from config.js, I would think doing a makeServer export would be in line. Keeping with clear point of where things are happening, the startMirage.js in this case should NOT use testConfig and the default export (base-config), the user should call those within makeServer.

Heres what I see

- setupMirage.js stays the same
- update app/initializers/ember-cli-mirage.js to register the new makeServer. The other two (base-config and test-config) would eventually be removed
- startMirage.js performs the same for backward compatibility but also detects the presence of a makeServer export. Instead of creating a new Server, calls said function that returns the new server created. Still pass the same options to makeServer that it was passing into new Server. baseConfig and testConfig will not be set to force an import or method call in makeServer for clarity (mentioned above)?
- the method getModels can be imported from util/ember-data, but I would rather it be re-exported from ember-cli-mirage to keep a clean import.

I think that covers it. Would you like to see a PR for this alone, and then a follow up PR that provides serializer discovery? Not sure I understand all the testing to create another app that does this way, Ill ask if I get into any problems.

ec-mirage documentation does not seem to be within this repo, how would I go about updating documentation?

**Comment by cah-briangantzler** [comment]
Awaiting your comments in case there are any changes. I have a PR ready to submit if the above is acceptable.

the users config.js can now look like this 
```
import { Server } from 'miragejs';
import { discoverEmberDataModels } from 'ember-cli-mirage';

export function makeServer(options) {
  options.routes = routeHandlers;
  discoverEmberDataModels(options);
  return new Server(options)
}

export function routeHandlers() {

  this.resource('user');

}
```
While the route handler can still be a default export if you wanted, it doesnt have to be.

**Comment by samselikoff** [comment]
Been busy so will have to check this soon. Would say hold off on writing any more code just in case there are changes.

(I know it’d be easier to have this convo in real-time but if it wasn’t async it wouldn’t be happening at all! This is the only way I can fit it into my schedule.)

**Comment by cah-briangantzler** [comment]
There was very little code. I am happy with the end result.

**Comment by samselikoff** [comment]
I appreciate your work! But I want to discuss the API before looking at the implementation. There's a lot of code in your PR... I don't know if we'll need an entire new test project for this. (Each one is just a lot to maintain.)

Can we keep discussing the API? I need to understand it before moving to review the code.

What is `routeHandlers` and how is it + `makeServer` used?

**Comment by cah-briangantzler** [comment]
Yea sure.  I wanted to show that the old and the new way both still work, so I created a new test project. Can do whatever you want there.

(sorry I left export on the routeHandlers, it isnt needed, which may have been your confusion)

Currently config.js contains an export default function which is the router handers, it just was not  named. 

```js
export default function() {
  this.resource('user');
}
```


We no longer need this function as an export default and we need it named to refer to it in the line `options.routes = routeHandlers`. (or you could choose to inline it, as mirageJS documentation shows. But as this grows, most people put it in a separate file and import it. We can now do this in ember-cli-mirage as well. examples below)   

Instructions to change config.js (simple straightforward solution)
- remove export default from the current function and give it a name
- add the imports and the makeServer function and assign `options.routes` the new name you assigned to the function (example below) 

```js
import { Server } from 'miragejs';
import { discoverEmberDataModels } from 'ember-cli-mirage';

export function makeServer(options) {
  options.routes = routeHandlers;
  discoverEmberDataModels(options);
  return new Server(options)
}

function routeHandlers() {
  this.resource('user');
}
```


Since the route handler function is now just a function, a user could arrive at the following more complex instuctions
- rename config.js to route-handers.js. (or any other name you want)
- create a new file named config.js that looks like the file below. Update the import statement of routeHandlers if you choose a different location or file name

file route-handlers.js (this can be any name the user wants).
```js
export default function() {
  this.resource('user');
}
```

config.js
```js
import { Server } from 'miragejs';
import { discoverEmberDataModels } from 'ember-cli-mirage';
import routeHanders from './route-handlers';  // where the user put it and what they called it

export function makeServer(options) {
  options.routes = routeHandlers;
  discoverEmberDataModels(options);
  return new Server(options)
}
```

**Comment by samselikoff** [comment]
Gotcha.

Ok first, just to minimize churn, I don't think we need to rename the default export to `routeHandlers`. I know it's a little confusing but at this point, the default function export from `config.js` has been the routes for so long that I don't think there's much of a benefit in changing it.

I also think they should already be attached to the config/options object passed into `makeServer`.

Basically, the default `makeServer` function is

```js
export function makeServer(config) {
  return new Server(config);
}
```

And then folks can augment the config as they please.

For `discoverEmberDataModel`, (1) I think it should return an object (instead of mutating the arg), second, does it need to be passed in the config at all?

I think ideally I'd like to see this:

```js
import { Server } from 'miragejs';
import { discoverEmberDataModels } from 'ember-cli-mirage';

export function makeServer(config) {
  let finalConfig = { ...config }
  finalConfig.models = discoverEmberDataModels();

  return new Server(finalConfig);
}

// as usual
export default function() {
  this.resource('users')
}
```

**Comment by cah-briangantzler** [comment]
So agree with all.  

Notes: :) 

I thought the removal of export default was part of the removal of magic to stear towards clearer what was happening. Both easily supported.

I went back and forth on discover models on passing the options or not (it returns the models so the assignment was available). I thought by passing the config, if additional config was needed,  it could be added with a hope of no additional changes on the users part. No problem to drop it.

End result we are on the same page and the current code I wrote needs like 2 or 3 lines tweaked and then we are there.


Documentation:
   I see that the documentation is in the dummy app (missed it behind pods). Would you like me to take a stab at altering the documentation in this PR, a separate followup PR, or would you prefer to do it (Im not the best doc writer but can try if you would like)


Out of curiosity, the config could have been altered directly, yet you created a copy as finalConfig. What was the through process behind that?   

I needed to read farther down the page on the documentation, I didnt realize the Spread operator could also be applied to properties, only ever seen it used on arrays.

**Comment by samselikoff** [comment]
> I thought the removal of export default was part of the removal of magic to stear towards clearer what was happening.

I think this requires more consideration/work since ultimately it would be a breaking change and to keep the scope of this as small as possible, I think we should have an additive API so you can get your original problem solved. Eventually we can figure out how to move ec-mirage community all the way over to an integration story that looks like every other framework but I don't think it's necessary for this change.

> I thought by passing the config, if additional config was needed, it could be added with a hope of no additional changes on the users part. No problem to drop it.

Yeah, I think the lesson I've learned from maintaining Mirage is when I try to do stuff like this it always ends up biting me. I think I'd rather have it do less + tell users to copy+paste more explicit code, since this is sort of a power user feature.

> Would you like me to take a stab at altering the documentation

Would be great as part of this PR, why don't you do a first pass + I can come clean it up as needed?

> Out of curiosity, the config could have been altered directly, yet you created a copy as finalConfig. What was the through process behind that?

Ya `config.models = models` is probably how I'd do it but, I think avoiding assignment/working with "immutable" data is a more popular/recognizable approach these days, so I did that just to make it clear that nothing special was going on and the user is free to alter what they pass in any way they like.

> I needed to read farther down the page on the documentation, I didnt realize the Spread operator could also be applied to properties, only ever seen it used on arrays.

Yep, especially having dipped my toes into React + Vue ecosystem over the past year this is very popular. So I think it'd be good to adopt the prominent style.

Thanks for continuing to push this forward Brian!

**Comment by cah-briangantzler** [comment]
the removal for the need of export default is ONLY if they create a makeServer function, export default for the old way is still supported. So I wouldnt consider that a breaking change since they already committed to a change by making a makeServer function. It also reinforces that is doesnt HAVE to be an export default in this file. They can now put the route handlers in a separate file(s) of their choosing and import it. (if they export default, we will continue to add it to the config, I would prefer to teach away from this and deprecate it at some point, rather the continue to support it).

That being said, any plan to deprecate the old way (I think it should, it leads to more code being removed from ec-mirage and the new code goes more in matching mirageJS documentation on how to define a server).

Testing.    The project 3 is huge I admit, but how do I prove the new way worked otherwise? Any ideas? If old way is deprecated, then test 1 would be removed along with it when the time comes.

**Comment by cah-briangantzler** [comment]
I went to change the discover models method and this is why I passed options in.     The discovered ED models has to be merged with the mirage models with the mirage models winning.   This line `options.models = Object.assign({}, getModels(), options.models);`.     

If we dont pass it in, the user would have to write this line
`options.models = Object.assign({}, discoverEmberDataModels(), options.models);`

not so optimal. Would prefer to hid that from them.

**Comment by cah-briangantzler** [comment]
Additional questions.

Should we use the `createServer` instead of `new Server` in the docs?

Would you like the blueprints updated for this? Keeping with the export default I assume. However what is the desired future for the config, do we use inline definition of routes and models as the MirageJS documentation does. 

```js
import { Server } from 'miragejs';
import { mergeEmberDataModels } from 'ember-cli-mirage';

export function makeServer(config) {
   let finalConfig = {
         ...config,
         routes() { 
              this.resource('users')
         },
        models: mergeEmberDataModels(config.models)
   };

  return new Server(finalConfig);
}
```

I realize it is simpler and easier for the docs than dealing with multiple files, should we also take this approach for documenting this way and the blueprint would look like
```js
import { Server } from 'miragejs';
import { mergeEmberDataModels } from 'ember-cli-mirage';

export function makeServer(config) {
   let finalConfig = {
         ...config,
         routes() { 
                // These comments are here to help you get started. Feel free to delete them.

                /*
                    Config (with defaults).

                    Note: these only affect routes defined *after* them!
                */

               // this.urlPrefix = '';    // make this `http://localhost:8080`, for example, if your API is on a different server
               // this.namespace = '';    // make this `/api`, for example, if your API is namespaced
               // this.timing = 400;      // delay for each request, automatically set to 0 during testing

               /*
                     Shorthand cheatsheet:

                     this.get('/posts');
                     this.post('/posts');
                     this.get('/posts/:id');
                     this.put('/posts/:id'); // or this.patch
                     this.del('/posts/:id');

                     https://www.ember-cli-mirage.com/docs/route-handlers/shorthands
              */         
        },
        // Remove or comment out the following line to remove Ember Data Model discovery and merge
        models: mergeEmberDataModels(config.models)
   };

  return new Server(finalConfig);
}
```

This has the benefit of looking like the MirageJS documentation

I wish the makeServer function was the unnamed export default from config, while being a breaking change, its a small change and better positions ec-mirage to more closely match the usage of mirageJS proper.

**Comment by cah-briangantzler** [comment]
If you could re-evaluate the PR https://github.com/miragejs/ember-cli-mirage/pull/1994 I would appreciate it.

**Comment by samselikoff** [comment]
Working on getting caught up on some things but this is definitely on my list. Appreciate the work!

– Sam

**Comment by cah-briangantzler** [comment]
Bump

**Comment by cah-briangantzler** [comment]
This issue got taken over with some pre-work to expose a makeServer option to allow better access to the underlying Mirage config.    Closing this issue and opening another one to allow for the serializer generation
